optimum/amd/ryzenai/modeling_decoder.py

```diff
@@ -0,0 +1,412 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+# Licensed under the MIT License.
+"""RyzenAIModelForCausalLM classes, allowing to run ONNX Models with ONNX Runtime VITIS-AI EP using the same API as Transformers."""
+
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import onnxruntime as ort
+import torch
+
+from optimum.utils import NormalizedConfigManager, check_if_transformers_greater
+from transformers import (
+    AutoModelForCausalLM,
+    GenerationConfig,
+)
+from transformers.modeling_outputs import CausalLMOutputWithPast
+
+from .modeling import RyzenAIModel
+from .utils import set_builtins, set_environment_variables
+
+
+if TYPE_CHECKING:
+    from transformers import PretrainedConfig
+
+if check_if_transformers_greater("4.25.0"):
+    from transformers.generation import GenerationMixin
+else:
+    from transformers.generation_utils import GenerationMixin
+
+
+class RyzenAIModelForCausalLM(RyzenAIModel, GenerationMixin):
+    """
+    Runs model with causal language modeling head using ONNX Runtime VITIS-AI EP.
+    """
+
+    main_input_name = "input_ids"
+    auto_model_class = AutoModelForCausalLM
+
+    def __init__(
+        self,
+        model: ort.InferenceSession,
+        config: "PretrainedConfig",
+        vaip_config: Union[str, Path] = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        preprocessors: Optional[List] = None,
+        generation_config: Optional[GenerationConfig] = None,
+        use_cache: Optional[bool] = None,
+        **kwargs,
+    ):
+        super().__init__(model, config, vaip_config, model_save_dir, preprocessors, **kwargs)
+
+        self._initialize_params(use_cache, generation_config)
+
+        self.use_fp16 = False
+        for inp in model.get_inputs():
+            if (
+                inp.name == "past_key_values" or inp.name in self.key_value_input_names
+            ) and inp.type == "tensor(float16)":
+                self.use_fp16 = True
+                break
```

Comment:
fp16 is not and never will be supported no? We could probably remove fp16 related code if so

Reply:
I did see some references of fp16 in their quantizer, but nothing concrete on device inference. So will remove it.